File: easy/3471-find-the-largest-almost-missing-integer/solution.py
from typing import List
from collections import defaultdict


class Solution:
    def largestInteger(self, nums: List[int], k: int) -> int:
        """
        3471. Find the Largest Almost Missing Integer
        Time: O(n * k)
        Space: O(n)
        """
        n = len(nums)
        windows = defaultdict(int)  # value -> number of size-k windows containing it

        for i in range(n - k + 1):
            # a value repeated inside one window still counts that window once
            for v in set(nums[i:i + k]):
                windows[v] += 1

        best = -1
        for v, c in windows.items():
            if c == 1 and v > best:
                best = v
        return best

File: easy/3471-find-the-largest-almost-missing-integer/solution.java
import java.util.HashSet;
import java.util.Set;

/**
 * 3471. Find the Largest Almost Missing Integer
 * Time: O(n * k)
 * Space: O(n)
 */
class Solution {
    public int largestInteger(int[] nums, int k) {
        int n = nums.length;
        int[] windows = new int[51]; // value -> number of size-k windows containing it

        for (int i = 0; i + k <= n; i++) {
            // a value repeated inside one window still counts that window once
            Set<Integer> seen = new HashSet<>();
            for (int j = i; j < i + k; j++) {
                seen.add(nums[j]);
            }
            for (int v : seen) {
                windows[v]++;
            }
        }

        for (int v = 50; v >= 0; v--) {
            if (windows[v] == 1) {
                return v;
            }
        }
        return -1;
    }
}
